Dia requires libxml2.

KineticTheory opened the pull request:
At least for me on RHEL7, Dia also depends on libxml2.

adamjstewart commented:
Can you run `ldd -r` on any of the dynamic executables or shared libraries that get installed and post the results here? It would be good to make sure it isn't linking to any other system dependencies that may or may not be installed.

junghans commented:
@adamjstewart: `scanelf -n` is actually the better option as `ldd -r`shows all loaded object incl. the one loaded by dependent libraries.

adamjstewart commented:
Ooh, didn't know about that one. I'll have to check it out. I think the equivalent on macOS is `otool -L`? If you know of a better option let me know. Maybe scanelf works there too.

junghans commented:
It seems to have `macos` keywords on [Gentoo](https://gitweb.gentoo.org/repo/gentoo.git/tree/app-misc/pax-utils/pax-utils-1.2.2.ebuild) so it should work there, too.

KineticTheory commented:
Looks okay to me. As expected, I'm picking up some X11 libraries, libc and libm.

```
% module load dia
% ldd -r `which dia`
	linux-vdso.so.1 =>  (0x00007ffc295ef000)
	libdia.so => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/dia-0.97.3-pdte26vizhmbwoczk3o3jsz4fcj32b26/lib/dia/libdia.so (0x00007f065ce1d000)
	libgtk-x11-2.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/gtkplus-2.24.31-c7x2eopbnsbpsu55e5loetv46nky6efo/lib/libgtk-x11-2.0.so.0 (0x00007f065c7ed000)
	libgdk-x11-2.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/gtkplus-2.24.31-c7x2eopbnsbpsu55e5loetv46nky6efo/lib/libgdk-x11-2.0.so.0 (0x00007f065c53e000)
	libatk-1.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/atk-2.20.0-4jxvwmka4zlpicnwvru7kf6t3fls6iu4/lib/libatk-1.0.so.0 (0x00007f065c318000)
	libgio-2.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/glib-2.49.7-t4wmsp6yzy6s4histek5bemxo2kmpogf/lib/libgio-2.0.so.0 (0x00007f065bf7a000)
	libpangocairo-1.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/pango-1.40.3-bzovwmx7salrttspsacuqdhdk3r2xe2v/lib/libpangocairo-1.0.so.0 (0x00007f065bd6c000)
	libcairo.so.2 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/cairo-1.14.8-bcip4nfy3kuxpzhsud6qhefzxjmnx4aq/lib/libcairo.so.2 (0x00007f065ba75000)
	libgdk_pixbuf-2.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/gdk-pixbuf-2.31.2-k6i7f5ooe5h3x5urqphdp6xywmt44c7f/lib/libgdk_pixbuf-2.0.so.0 (0x00007f065b851000)
	libgthread-2.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/glib-2.49.7-t4wmsp6yzy6s4histek5bemxo2kmpogf/lib/libgthread-2.0.so.0 (0x00007f065b64e000)
	libpangoft2-1.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/pango-1.40.3-bzovwmx7salrttspsacuqdhdk3r2xe2v/lib/libpangoft2-1.0.so.0 (0x00007f065b437000)
	libpango-1.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/pango-1.40.3-bzovwmx7salrttspsacuqdhdk3r2xe2v/lib/libpango-1.0.so.0 (0x00007f065b1ec000)
	libfontconfig.so.1 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/fontconfig-2.11.1-h3l7yisyznyemlfrv6egzgvx47tyb2ou/lib/libfontconfig.so.1 (0x00007f065afae000)
	libgobject-2.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/glib-2.49.7-t4wmsp6yzy6s4histek5bemxo2kmpogf/lib/libgobject-2.0.so.0 (0x00007f065ad5d000)
	libglib-2.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/glib-2.49.7-t4wmsp6yzy6s4histek5bemxo2kmpogf/lib/libglib-2.0.so.0 (0x00007f065aa4c000)
	libfreetype.so.6 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/freetype-2.7-bufykqx7zxx6gbekpdmdmqbtw4ecbiwh/lib/libfreetype.so.6 (0x00007f065a7b3000)
	libintl.so.8 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/gettext-0.19.8.1-meti2rz4df6sr5auemuiexlq5enrmw3y/lib/libintl.so.8 (0x00007f065a5a7000)
	libxml2.so.2 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/libxml2-2.9.4-hvr44mn3lfne4352n42yem7vlepbf4bu/lib/libxml2.so.2 (0x00007f065a23e000)
	liblzma.so.5 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/xz-5.2.3-yxhjznhbx2nud4c5pxnsh4wsvwyzfd7v/lib/liblzma.so.5 (0x00007f065a018000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f0659df7000)
	libpng16.so.16 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/libpng-1.6.27-tsms3us3pbki4r73jqdlctifia3swg2h/lib/libpng16.so.16 (0x00007f0659bc5000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f06598c3000)
	libz.so.1 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/zlib-1.2.11-hpdicdxfq5jn2ycttiuiu42hemg444nf/lib/libz.so.1 (0x00007f06596a4000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f0659488000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f06590c7000)
	libXinerama.so.1 => /lib64/libXinerama.so.1 (0x00007f0658ec2000)
	libpixman-1.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/pixman-0.34.0-thnwhzrsh3xl4eehda3h4dubbvt7f5dp/lib/libpixman-1.so.0 (0x00007f0658c1b000)
	libXrender.so.1 => /lib64/libXrender.so.1 (0x00007f0658a10000)
	libSM.so.6 => /lib64/libSM.so.6 (0x00007f0658808000)
	libICE.so.6 => /lib64/libICE.so.6 (0x00007f06585ec000)
	libX11.so.6 => /lib64/libX11.so.6 (0x00007f06582ad000)
	libXext.so.6 => /lib64/libXext.so.6 (0x00007f065809b000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f0657e93000)
	libgmodule-2.0.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/glib-2.49.7-t4wmsp6yzy6s4histek5bemxo2kmpogf/lib/libgmodule-2.0.so.0 (0x00007f0657c8e000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f0657a74000)
	libharfbuzz.so.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/harfbuzz-0.9.37-jil4wd7ppg3w2u5dpltd3yv2wyjgbkvb/lib/libharfbuzz.so.0 (0x00007f065781d000)
	libffi.so.6 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/libffi-3.2.1-4s3dshkk5mra3bp2vdxqfasds2sxmzrp/lib64/libffi.so.6 (0x00007f0657612000)
	libpcre.so.1 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/pcre-8.40-3jnagmpqhzjxihvnvfswurbvqwoohsyg/lib/libpcre.so.1 (0x00007f06573cf000)
	libbz2.so.1.0 => /scratch/vendors/spack.20170420/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/bzip2-1.0.6-qrd3gzlkqpof6jthkmt2vejxvmbd3x3s/lib/libbz2.so.1.0 (0x00007f06571be000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f065d0ae000)
	libuuid.so.1 => /lib64/libuuid.so.1 (0x00007f0656fb8000)
	libxcb.so.1 => /lib64/libxcb.so.1 (0x00007f0656d95000)
	libXau.so.6 => /lib64/libXau.so.6 (0x00007f0656b91000)
```

adamjstewart commented:
Can you add dependencies on all of the X11 libs as well? I added packages for everything on Xorg once upon a time.

adamjstewart commented:
Just check to make sure that it's actually dia that needs them and not one of its deps.

KineticTheory commented:
@adamjstewart I'm not clear on your concern.  However, it was easy for me to do as you requested.  I added the following dependencies to dia/package.py:

```
    depends_on('libice')
    depends_on('libsm')
    depends_on('libuuid')
    depends_on('libx11')
    depends_on('libxext')
    depends_on('libxinerama')
    depends_on('libxrender')
```

This generates the following specification for dia:

```
Concretized
--------------------------------
dia@0.97.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^cairo@1.14.8%gcc@4.8.5~X arch=linux-rhel7-x86_64 
        ^fontconfig@2.11.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^font-util@1.3.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
                ^pkg-config@0.29.2%gcc@4.8.5+internal_glib arch=linux-rhel7-x86_64 
                ^util-macros@1.19.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^freetype@2.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
                ^bzip2@1.0.6%gcc@4.8.5+shared arch=linux-rhel7-x86_64 
                ^libpng@1.6.27%gcc@4.8.5 arch=linux-rhel7-x86_64 
                    ^zlib@1.2.11%gcc@4.8.5+pic+shared arch=linux-rhel7-x86_64 
            ^libxml2@2.9.4%gcc@4.8.5~python arch=linux-rhel7-x86_64 
                ^xz@5.2.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^glib@2.49.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^autoconf@2.69%gcc@4.8.5 arch=linux-rhel7-x86_64 
                ^m4@1.4.18%gcc@4.8.5+sigsegv arch=linux-rhel7-x86_64 
                    ^libsigsegv@2.11%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^automake@1.15%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^gettext@0.19.8.1%gcc@4.8.5+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-rhel7-x86_64 
                ^ncurses@6.0%gcc@4.8.5~symlinks arch=linux-rhel7-x86_64 
                ^tar@1.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^libffi@3.2.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^libtool@2.4.6%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^pcre@8.40%gcc@4.8.5+utf arch=linux-rhel7-x86_64 
        ^pixman@0.34.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^gtkplus@2.24.31%gcc@4.8.5~X arch=linux-rhel7-x86_64 
        ^atk@2.20.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^gdk-pixbuf@2.31.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^jpeg@9b%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^libtiff@4.0.6%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^pango@1.40.3%gcc@4.8.5~X arch=linux-rhel7-x86_64 
            ^harfbuzz@0.9.37%gcc@4.8.5 arch=linux-rhel7-x86_64 
                ^icu4c@58.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^intltool@0.51.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^libice@1.0.9%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^xproto@7.0.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^xtrans@1.3.5%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^libsm@1.2.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^libuuid@1.0.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^libx11@1.6.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^inputproto@2.3.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^kbproto@1.0.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^libxcb@1.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^libpthread-stubs@0.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^libxau@1.0.8%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^libxdmcp@1.1.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^python@2.7.13%gcc@4.8.5~tk~ucs4 arch=linux-rhel7-x86_64 
                ^openssl@1.0.2k%gcc@4.8.5 arch=linux-rhel7-x86_64 
                ^readline@7.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
                ^sqlite@3.18.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^xcb-proto@1.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^xextproto@7.3.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^libxext@1.3.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^libxinerama@1.1.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^xineramaproto@1.2.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^libxrender@0.9.9%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^renderproto@0.11.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^libxslt@1.1.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
        ^libgcrypt@1.6.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
            ^libgpg-error@1.21%gcc@4.8.5 arch=linux-rhel7-x86_64 
    ^swig@3.0.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
```
The remaining non-spack dependencies are:

```
ldd -r /scratch/vendors/spack/.../bin/dia | grep -v spack
	linux-vdso.so.1 =>  (0x00007ffc56b6d000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007efbf9dbf000)
	libm.so.6 => /lib64/libm.so.6 (0x00007efbf988a000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007efbf944f000)
	libc.so.6 => /lib64/libc.so.6 (0x00007efbf908e000)
	librt.so.1 => /lib64/librt.so.1 (0x00007efbf7e5c000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007efbf7a3e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007efbfd076000)
```

Is there anything in this list that concerns you?  Do you want me to add these X11 dependencies to `dia/package.py` as part of this PR?

adamjstewart commented:
I don't see anything else that's missing.

In general (not just for dia), every Spack package should declare every dependency it can. It's not hard to foresee a user trying to build dia on a machine without any X11 libraries. Making sure all of the dependencies are declared is important for reliable, reproducible, and stable builds.

Just make sure that it's actually dia that depends on these libraries and not one of it's dependencies. If you run `ldd -r` on all of its dependencies and none of them link to X11 libraries, then it's safe to add them to dia.

KineticTheory commented:
@adamjstewart I added the X11 dependencies and I cleaned up the dependency list.  According to the dependency diagram below, Dia shouldn't need an explicit `depends_on('libxml2')`, but I couldn't get it to build without it.
![dia](https://cloud.githubusercontent.com/assets/17579601/25403143/9c8666a6-29b8-11e7-8a05-eb0b07eabc16.png)

adamjstewart commented on var/spack/repos/builtin/packages/dia/package.py:
```diff
@@ -34,15 +34,18 @@ class Dia(Package):
     depends_on('intltool', type='build')
     depends_on('gtkplus@2.6.0:')
-    depends_on('cairo')
-    depends_on('libpng')
```
You're sure these dependencies aren't needed? They might link to a different executable or library. I'm not familiar with dia though.

KineticTheory replied:
My understanding is that cairo is a transitive dependency: 
```dia -> gtkplus -> pango -> cairo```
Similarly for libpng:
```dia -> gtkplus -> gdk-pixbuf -> libpng```

adamjstewart replied:
See below. If `dia` depends on `cairo` and `libpng`, it should declare it as a direct dependency. We shouldn't rely on transitive dependencies to always be in RPATH.

davydden replied:
To be sure, you can check in the build logs of `dia` if `cairo` or `libpng` are mentioned anyhow.

davydden replied:
if it is needed by `dia`, then, indeed, we shall not rely on them being in `RPATH` and should keep these `depends_on()`.

adamjstewart commented:
@scheibelp Any idea why `libxml2` wouldn't be present as a transitive dependency? Maybe this package needs it as a build dependency and we are only adding it as a link dependency?

scheibelp commented:
> @scheibelp Any idea why libxml2 wouldn't be present as a transitive dependency?

It appears that it is present as a transitive dependency (I see it in the diagram that @KineticTheory added here). Note that not all dependencies are treated the same though; just showing up as a transitive dependency may not be sufficient because build_environment may set up the build_environment differently for transitive dependencies depending on their dep type.

@KineticTheory if you post the result of ```spack spec -t dia``` and the error message, I could explain more exactly how Spack is setting up the build environment WRT the transitive libxml2. In Some cases you will want to add libxml2 as a direct dependency even if it would otherwise appear as a transitive dependency (the error message may help me figure out whether that's true here).

adamjstewart commented:
In general, if a package depends on something, it should _directly_ depend on something. We shouldn't rely on indirect dependencies. Once upon a time, @davydden (I think?) had a problem building `dealii` because there were too many dependencies in RPATH. We made a change that had the affect of only _direct_ dependencies showing up in RPATH and quickly found that half of the packages in Spack broke. If we ever do go back down that route (not a terrible idea) we will want to make sure that every package correctly lists its dependencies.

davydden commented:
@adamjstewart indeed, i had issues on macOS due to too many load commands.
FYI, this behaviour is now optional via `transitive_rpaths = False` member variable in `Package`.

adamjstewart commented:
For what it's worth, I would be in favor of going back to only adding direct dependencies to RPATH **after we have proper package testing to see what breaks**.

KineticTheory commented:
@scheibelp Here's the specification:
```
 % spack spec -t dia%gcc@4.8.5

Normalized
--------------------------------
[   ]  dia%gcc@4.8.5
[bl ]      ^gtkplus@2.6.0:
[bl ]          ^atk
[bl ]              ^glib
[b  ]                  ^autoconf
[b r]                      ^m4@1.4.6:
[b  ]                  ^automake
[bl ]                  ^gettext
[bl ]                  ^libffi
[b  ]                  ^libtool
[b  ]                  ^pkg-config@0.9.0:+internal_glib
[bl ]                  ^zlib@1.0.4:
[bl ]          ^gdk-pixbuf
[bl ]              ^jpeg
[bl ]              ^libpng
[bl ]              ^libtiff
[bl ]                  ^xz
[bl ]          ^pango
[bl ]              ^cairo
[bl ]                  ^fontconfig@2.10.91:
[bl ]                      ^font-util
[b  ]                          ^util-macros
[bl ]                      ^freetype
[bl ]                          ^bzip2
[bl ]                      ^libxml2
[bl ]                  ^pixman
[bl ]              ^harfbuzz
[bl ]                  ^icu4c
[b  ]      ^intltool
[bl ]      ^libsm
[bl ]          ^libice@1.0.5:
[b  ]              ^xproto@7.0.17:
[b  ]              ^xtrans
[bl ]      ^libuuid
[bl ]      ^libxinerama
[bl ]          ^libx11@1.6:
[b  ]              ^inputproto
[b  ]              ^kbproto
[bl ]              ^libxcb@1.1.92:
[bl ]                  ^libpthread-stubs
[bl ]                  ^libxau@0.99.2:
[bl ]                  ^libxdmcp
[b  ]                  ^python@2:2.8
[bl ]                      ^ncurses
[bl ]                      ^openssl
[bl ]                      ^readline
[bl ]                      ^sqlite
[b  ]                  ^xcb-proto
[b  ]              ^xextproto@7.1.99:
[bl ]          ^libxext
[b  ]          ^xineramaproto@1.1.99.1:
[bl ]      ^libxrender
[b  ]          ^renderproto@0.9:
[bl ]      ^libxslt
[bl ]          ^libgcrypt
[bl ]              ^libgpg-error
[bl ]      ^swig
[bl ]          ^pcre

Concretized
--------------------------------
[   ]  dia@0.97.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^gtkplus@2.24.31%gcc@4.8.5~X arch=linux-rhel7-x86_64 
[bl ]          ^atk@2.20.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^glib@2.49.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^autoconf@2.69%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b r]                      ^m4@1.4.18%gcc@4.8.5+sigsegv arch=linux-rhel7-x86_64 
[bl ]                          ^libsigsegv@2.11%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^automake@1.15%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^gettext@0.19.8.1%gcc@4.8.5+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-rhel7-x86_64 
[bl ]                      ^bzip2@1.0.6%gcc@4.8.5+shared arch=linux-rhel7-x86_64 
[bl ]                      ^libxml2@2.9.4%gcc@4.8.5~python arch=linux-rhel7-x86_64 
[b  ]                          ^pkg-config@0.29.2%gcc@4.8.5+internal_glib arch=linux-rhel7-x86_64 
[bl ]                          ^xz@5.2.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                          ^zlib@1.2.11%gcc@4.8.5+pic+shared arch=linux-rhel7-x86_64 
[bl ]                      ^ncurses@6.0%gcc@4.8.5~symlinks arch=linux-rhel7-x86_64 
[bl ]                      ^tar@1.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^libffi@3.2.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^libtool@2.4.6%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^pcre@8.40%gcc@4.8.5+utf arch=linux-rhel7-x86_64 
[bl ]          ^gdk-pixbuf@2.31.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^jpeg@9b%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^libpng@1.6.27%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^libtiff@4.0.6%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^pango@1.40.3%gcc@4.8.5~X arch=linux-rhel7-x86_64 
[bl ]              ^cairo@1.14.8%gcc@4.8.5~X arch=linux-rhel7-x86_64 
[bl ]                  ^fontconfig@2.11.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                      ^font-util@1.3.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                          ^util-macros@1.19.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                      ^freetype@2.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^pixman@0.34.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^harfbuzz@0.9.37%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^icu4c@58.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]      ^intltool@0.51.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libsm@1.2.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^libice@1.0.9%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^xproto@7.0.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^xtrans@1.3.5%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libuuid@1.0.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libxinerama@1.1.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^libx11@1.6.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^inputproto@2.3.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^kbproto@1.0.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^libxcb@1.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^libpthread-stubs@0.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^libxau@1.0.8%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^libxdmcp@1.1.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^python@2.7.13%gcc@4.8.5~tk~ucs4 arch=linux-rhel7-x86_64 
[bl ]                      ^openssl@1.0.2k%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                      ^readline@7.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                      ^sqlite@3.18.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^xcb-proto@1.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^xextproto@7.3.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^libxext@1.3.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]          ^xineramaproto@1.2.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libxrender@0.9.9%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]          ^renderproto@0.11.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libxslt@1.1.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^libgcrypt@1.6.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^libgpg-error@1.21%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^swig@3.0.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
```

scheibelp commented:
Oh and to make sure this is for the tree where the build fails?

Spack adds direct build dependencies to PATH during a build, and all transitive link dependencies are RPATHed and also get added to includes. If this is the DAG where an error occurs I'm guessing the issue is that it was not listed as a direct build dependency so whatever binaries are provided were not available.

If dia needs libxml2 because a build dependency runs an exe it provides, then IMO it shouldn't have to be listed as a direct dependency (although to be clear at the moment this is required). There are a couple PR's which aim to fix this, e.g. https://github.com/LLNL/spack/pull/3347. And also a more general discussion about how deptypes should be handled at https://github.com/LLNL/spack/issues/3768

junghans commented:
dia depends on libxml2 directly: https://git.gnome.org/browse/dia/tree/lib/sheet.c#n30

KineticTheory commented:
@scheibelp Oh - sorry for the misunderstanding.  Here's the spec for the failing build (but as @junghans mentioned, it appears there is a direct dependency):
```
% spack spec -t dia%gcc@4.8.5
Input spec
--------------------------------
[   ]  dia%gcc@4.8.5

Normalized
--------------------------------
[   ]  dia%gcc@4.8.5
[bl ]      ^gtkplus@2.6.0:
[bl ]          ^atk
[bl ]              ^glib
[b  ]                  ^autoconf
[b r]                      ^m4@1.4.6:
[b  ]                  ^automake
[bl ]                  ^gettext
[bl ]                  ^libffi
[b  ]                  ^libtool
[b  ]                  ^pkg-config@0.9.0:+internal_glib
[bl ]                  ^zlib@1.0.4:
[bl ]          ^gdk-pixbuf
[bl ]              ^jpeg
[bl ]              ^libpng
[bl ]              ^libtiff
[bl ]                  ^xz
[bl ]          ^pango
[bl ]              ^cairo
[bl ]                  ^fontconfig@2.10.91:
[bl ]                      ^font-util
[b  ]                          ^util-macros
[bl ]                      ^freetype
[bl ]                          ^bzip2
[bl ]                      ^libxml2
[bl ]                  ^pixman
[bl ]              ^harfbuzz
[bl ]                  ^icu4c
[b  ]      ^intltool
[bl ]      ^libsm
[bl ]          ^libice@1.0.5:
[b  ]              ^xproto@7.0.17:
[b  ]              ^xtrans
[bl ]      ^libuuid
[bl ]      ^libxinerama
[bl ]          ^libx11@1.6:
[b  ]              ^inputproto
[b  ]              ^kbproto
[bl ]              ^libxcb@1.1.92:
[bl ]                  ^libpthread-stubs
[bl ]                  ^libxau@0.99.2:
[bl ]                  ^libxdmcp
[b  ]                  ^python@2:2.8
[bl ]                      ^ncurses
[bl ]                      ^openssl
[bl ]                      ^readline
[bl ]                      ^sqlite
[b  ]                  ^xcb-proto
[b  ]              ^xextproto@7.1.99:
[bl ]          ^libxext
[b  ]          ^xineramaproto@1.1.99.1:
[bl ]      ^libxrender
[b  ]          ^renderproto@0.9:
[bl ]      ^libxslt
[bl ]          ^libgcrypt
[bl ]              ^libgpg-error
[bl ]      ^swig
[bl ]          ^pcre

Concretized
--------------------------------
[   ]  dia@0.97.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^gtkplus@2.24.31%gcc@4.8.5~X arch=linux-rhel7-x86_64 
[bl ]          ^atk@2.20.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^glib@2.49.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^autoconf@2.69%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b r]                      ^m4@1.4.18%gcc@4.8.5+sigsegv arch=linux-rhel7-x86_64 
[bl ]                          ^libsigsegv@2.11%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^automake@1.15%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^gettext@0.19.8.1%gcc@4.8.5+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-rhel7-x86_64 
[bl ]                      ^bzip2@1.0.6%gcc@4.8.5+shared arch=linux-rhel7-x86_64 
[bl ]                      ^libxml2@2.9.4%gcc@4.8.5~python arch=linux-rhel7-x86_64 
[b  ]                          ^pkg-config@0.29.2%gcc@4.8.5+internal_glib arch=linux-rhel7-x86_64 
[bl ]                          ^xz@5.2.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                          ^zlib@1.2.11%gcc@4.8.5+pic+shared arch=linux-rhel7-x86_64 
[bl ]                      ^ncurses@6.0%gcc@4.8.5~symlinks arch=linux-rhel7-x86_64 
[bl ]                      ^tar@1.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^libffi@3.2.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^libtool@2.4.6%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^pcre@8.40%gcc@4.8.5+utf arch=linux-rhel7-x86_64 
[bl ]          ^gdk-pixbuf@2.31.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^jpeg@9b%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^libpng@1.6.27%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^libtiff@4.0.6%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^pango@1.40.3%gcc@4.8.5~X arch=linux-rhel7-x86_64 
[bl ]              ^cairo@1.14.8%gcc@4.8.5~X arch=linux-rhel7-x86_64 
[bl ]                  ^fontconfig@2.11.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                      ^font-util@1.3.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                          ^util-macros@1.19.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                      ^freetype@2.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^pixman@0.34.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^harfbuzz@0.9.37%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^icu4c@58.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]      ^intltool@0.51.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libsm@1.2.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^libice@1.0.9%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^xproto@7.0.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^xtrans@1.3.5%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libuuid@1.0.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libxinerama@1.1.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^libx11@1.6.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^inputproto@2.3.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^kbproto@1.0.7%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^libxcb@1.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^libpthread-stubs@0.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^libxau@1.0.8%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                  ^libxdmcp@1.1.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^python@2.7.13%gcc@4.8.5~tk~ucs4 arch=linux-rhel7-x86_64 
[bl ]                      ^openssl@1.0.2k%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                      ^readline@7.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]                      ^sqlite@3.18.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]                  ^xcb-proto@1.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]              ^xextproto@7.3.0%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^libxext@1.3.3%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]          ^xineramaproto@1.2.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libxrender@0.9.9%gcc@4.8.5 arch=linux-rhel7-x86_64 
[b  ]          ^renderproto@0.11.1%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^libxslt@1.1.29%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]          ^libgcrypt@1.6.2%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]              ^libgpg-error@1.21%gcc@4.8.5 arch=linux-rhel7-x86_64 
[bl ]      ^swig@3.0.12%gcc@4.8.5 arch=linux-rhel7-x86_64 
```

scheibelp commented:
Thanks for the update! I was not specific when I asked. In this case my additions confuse the issue since I was speaking generally. @junghans demonstrates it should be a direct dependency in this case (and using ```depends_on(libxml2)``` - which by default creates a build+link dependency - sounds like the way to go).

hartzell commented:
> It's not hard to foresee a user trying to build dia on a machine without any X11 libraries

I just gave this a try on a CentOS box with nearly no X libraries (someone added a couple...) and the `gtk+` build failed with 

```
checking for CAIRO_BACKEND... no
configure: error: Package requirements (cairo-xlib >= 1.6) were not met:
```

I'll try to poke at this a bit more in the next day or two but may not be able to spend much time on it.

tgamblin commented:
Seems like we should add a `depends_on('cairo@1.6:')` to `gtkplus` -- does that work for you @hartzell?

KineticTheory commented:
@hartzell If @tgamblin suggestion works for you, I can add this change to #3976. Please let me know what you find.

hartzell commented:
@tgamblin -- that's not the issue, the change didn't fix anything.  The test that's failing seems to be about the cairo backend, not cairo itself.

I think (build in progress) that dia needs to depend on gtkplus+X so that the things downstream (including Cairo's set of backends) include X support.

`spack spec` looked ok with this change.  I need to remind myself about whether Cairo needs the same 3 depends_on dance that pango uses in the gtkplus package.  @adamjstewart has explained it before but I can't remember if the situation has changed since the last time I hit it.

hartzell commented:
I was afraid this was going to happen....

If I make add a `depends_on('gtkplus+X')` to dia then the build of gtkplus fails with:

```
ImportError: No module named site
make[4]: *** [Gdk-2.0.gir] Error 1
make[4]: Leaving directory `/tmp/hartzelg/spack-stage/spack-stage-QHEpfg/gtk+-2.24.31/gdk'
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory `/tmp/hartzelg/spack-stage/spack-stage-QHEpfg/gtk+-2.24.31/gdk'
make[2]: *** [all] Error 2
make[2]: Leaving directory `/tmp/hartzelg/spack-stage/spack-stage-QHEpfg/gtk+-2.24.31/gdk'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/tmp/hartzelg/spack-stage/spack-stage-QHEpfg/gtk+-2.24.31'
make: *** [all] Error 2
```

#3077 broke gtkplus (for me, at least).  I described this problem and am working towards a fix in #3208 but I wasn't happy with it and have been letting it stew a bit.

I'll spend some time on it today.

tgamblin commented:
@hartzell: I think we can merge this and do the other fixes in a separate PR., since this at least fixes RHEL7.

hartzell commented:
@tgamblin -- I don't think that this will work as merged.

`url_for_version` has a typo in the path to fetch the tarball:

```diff
@@ -49,7 +49,7 @@ class Dia(Package):

     def url_for_version(self, version):
         """Handle Dia's version-based custom URLs."""
-        return 'https://ftp.gnome.org/pub/gnome/source/dia/%s/dia-%s.tar.xz' % (version.up_to(2), version)
+        return 'https://ftp.gnome.org/pub/gnome/sources/dia/%s/dia-%s.tar.xz' % (version.up_to(2), version)
```

adamjstewart commented:
Looks like that was present before this PR. Want to submit a new PR?

hartzell commented:
Will do.  Also seems to be missing a dependency on freetype.  Building, building, building....

adamjstewart commented:
It never ends, does it.
